### ISSUE TITLE
large memory usage in io.grpc.netty.NettyClientTransport$2

### ISSUE THREAD

**Issue by garrettjonesgoogle** [grpc/grpc-java#2554]
As reported in https://github.com/GoogleCloudPlatform/google-cloud-java/issues/1449 , io.grpc.netty.NettyClientTransport$2 (I don't know which anonymous class it is) consumes a large amount of memory. Detailed memory analysis is contained in the linked issue. (This is blocking google-cloud-logging GA).

**Comment by carl-mastrangelo** [comment]
What version of gRPC?  Running at head shows (likely incorrectly):

```
        new NettyClientStream.TransportState(handler, maxMessageSize, statsTraceCtx) {
          @Override
          protected Status statusFromFailedFuture(ChannelFuture f) {
            return NettyClientTransport.this.statusFromFailedFuture(f);
          }
        }
```

**Comment by garrettjonesgoogle** [comment]
version `1.0.1`.

**Comment by ejona86** [comment]
Try updating to v1.0.2. In v1.0.2 we bumped the version of Netty to avoid a memory issue.

**Comment by garrettjonesgoogle** [comment]
As noted on the issue on google-cloud-java, they tried both Netty 4.1.3 and 4.1.5. What is supposed to be the version with fixed memory issues?

**Comment by ejona86** [comment]
The memory issue in Netty was similar to a leak, so it would only fix the "base memory usage".

`NettyClientTransport$2` is probably `NettyClientStream`. Both it and StreamBufferingEncoder may buffer outgoing data.

I dug in to cloud logging but couldn't find the code likely being tested. LocalLoggingImpl looks close, but seems like it is for testing. What code sends the RPCs? Is it observing flow control?

**Comment by garrettjonesgoogle** [comment]
Partial stack:

https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-logging/src/main/java/com/google/cloud/logging/spi/v2/LoggingServiceV2Client.java#L305

https://github.com/googleapis/gax-java/blob/master/src/main/java/com/google/api/gax/grpc/UnaryCallable.java#L284

https://github.com/googleapis/gax-java/blob/master/src/main/java/com/google/api/gax/grpc/DirectCallable.java#L55

(There is more stuff in between)

There is no explicit flow control being done in the stack.

**Comment by ejona86** [comment]
Okay, it looks like they are unary calls, so flow control isn't in the picture as much. After digging through a lot of delegation plumbing, it appears `LoggingHandler` is the class that actually chooses what/when to send an RPC.

And it looks like the default [flush size](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/628b7c931a82a40ac5978b1d75ca3fd79eb60da8/google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingHandler.java#L144) is "1". So after [every log statement](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/628b7c931a82a40ac5978b1d75ca3fd79eb60da8/google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingHandler.java#L290) it does a separate RPC.

There's things gRPC could do better, but doing an RPC each time you want to send 9 bytes will generate garbage.

**Comment by garrettjonesgoogle** [comment]
@gregw , what do you think? Should we bump the flush size up for logging?

**Comment by gregw** [comment]
@ejona86 @garrettjonesgoogle 

TL;DR, writing logs in bigger batches is certainly acceptable, but flush size is a poor mechanism to achieve that.

Let's say we increase it to 100, then we could get a lot of log messages sitting in the queue that are not flushed on a server with less verbose logging.   But then on a very verbose server, it could produce 1000s of log records per second and thus still fill up a queue if the network latency is high.

The flush mechanism probably should be a bit smarter, either:
 - have a timer that would send the buffer of logrecords even if less that the flush limit if it has been waiting some time, this would allow a large flush limit to be set.
 - alternately a limit on the number of outstanding RPC calls could be set (default 1), so that even if the flush limit is exceeded, if there are outstanding RPC calls, then the buffer queue will continue to grow. Then once a prior RPC completes, then entire queue is sent.

I prefer the second option, but the write async API only returns a `Future`, rather than a `CompletableFuture` so it is hard to implement efficiently, or at least without allocating a flushing thread.  But then having 1 (or a configurable few) thread dedicated to flushing logs is probably an acceptable cost - they could use the blocking API rather than async.

**Comment by garrettjonesgoogle** [comment]
Well we're in luck! We already have support in gax-java for something called "bundling", which supports having simultaneous limits on the count of objects to send, the memory size of the objects, and a time delay. I already configured it for logging, but the flush size is sitting on top of it currently. If we remove the flush size, we can depend on the bundling.

**Comment by gregw** [comment]
make it so!

So should we close here and discuss/test this in https://github.com/GoogleCloudPlatform/google-cloud-java/issues/1449 ?

**Comment by garrettjonesgoogle** [comment]
Sure, closing this issue.
